components/organisms/Chat/ChatMessages/MessageComponent.tsx

```diff
@@ -0,0 +1,195 @@
+import { Message } from "ai";
+import { CodeBlock } from "./CodeBlock";
+import React, { useContext, useEffect, useState } from "react";
+import { SoqlResult } from "./SoqlResult";
+
+import { ChatContext } from "./ChatContext";
+import { AiOutlineCopy } from "react-icons/ai";
+import { BiDownload } from "react-icons/bi";
+import { saveAs } from "file-saver";
+import Papa from "papaparse";
+import va from "@vercel/analytics";
+import { SalesforceQueryResultWithError } from "@/shared/types/salesforceTypes";
+
+type MessageProps = {
+  message: Message;
+};
+
+const cleanCode = (part: string) => {
+  let code = part.slice(3, -3);
+  if (code.startsWith("\n")) {
+    code = code.slice(1);
+  }
+
+  if (
+    code.startsWith("SOQL") ||
+    code.startsWith("soql") ||
+    code.startsWith("sql") ||
+    code.startsWith("bash")
+  ) {
+    code = code.slice(4);
+  }
+  if (code.startsWith("\n")) {
+    code = code.slice(1);
+  }
+  return code;
+};
+
+export const MessageComponent = ({ message }: MessageProps) => {
+  const parts = message?.content.split(/(```[\s\S]*?```)/gm);
+  const [queryResult, setQueryResult] =
+    useState<SalesforceQueryResultWithError>(
+      {} as SalesforceQueryResultWithError,
+    );
+  const { messages } = useContext(ChatContext);
+
+  const [saveLoading, setSaveLoading] = useState(false);
+  const [savedUrl, setSavedUrl] = useState(false);
+  const [downloadedUrl, setDownloadedUrl] = useState(false);
+
+  // USE EFFECTS TO MANAGE LOAD/DOWNLOAD STATE
```

Comment:
Why

Reply:
Just purely frontend-> to change the text to loaded after finished loading, then reset back to the original value after a bit

Reply:
Abstract into a hook so not repeating this twice